pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,505 @@
+			Some(proof) => proof,
+		}
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	pub(crate) fn upsert_pending_entry(
+		message_proof: Proof,
+		router_id: &T::RouterId,
+		new_inbound_entry: InboundEntry<T>,
+	) -> DispatchResult {
+		PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+			match storage_entry {
+				None => {
+					*storage_entry = Some(new_inbound_entry);
+
+					Ok::<(), DispatchError>(())
+				}
+				Some(stored_inbound_entry) => {
+					stored_inbound_entry.pre_dispatch_update(new_inbound_entry)
+				}
+			}
+		})
+	}
+
+	/// Checks if the number of proofs required for executing one message
+	/// were received, and if so, decreases the counts accordingly and executes
+	/// the message.
+	pub(crate) fn execute_if_requirements_are_met(
+		message_proof: Proof,
+		router_ids: &[T::RouterId],
+		session_id: T::SessionId,
+		expected_proof_count: u32,
+		domain_address: DomainAddress,
+	) -> DispatchResult {
+		let mut message = None;
+		let mut votes = 0;
+
+		for router_id in router_ids {
+			match PendingInboundEntries::<T>::get(message_proof, router_id) {
+				// We expected one InboundEntry for each router, if that's not the case,
+				// we can return.
+				None => return Ok(()),
+				Some(stored_inbound_entry) => match stored_inbound_entry {
+					InboundEntry::Message(message_entry) => message = Some(message_entry.message),
+					InboundEntry::Proof(proof_entry) => {
+						if proof_entry.has_valid_vote_for_session(session_id) {
+							votes.ensure_add_assign(1)?;
+						}
+					}
+				},
+			};
+		}
+
+		if votes < expected_proof_count {
+			return Ok(());
+		}
+
+		match message {
+			Some(msg) => {
+				Self::execute_post_voting_dispatch(
+					message_proof,
+					router_ids,
+					session_id,
+					expected_proof_count,
+				)?;
+
+				T::InboundMessageHandler::handle(domain_address, msg)
+			}
+			None => Ok(()),
+		}
+	}
+
+	/// Decreases the counts for inbound entries and removes them if the
+	/// counts reach 0.
+	pub(crate) fn execute_post_voting_dispatch(
+		message_proof: Proof,
+		router_ids: &[T::RouterId],
+		session_id: T::SessionId,
+		expected_proof_count: u32,
+	) -> DispatchResult {
+		for router_id in router_ids {
+			PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+				match storage_entry {
+					None => {
+						Err::<(), DispatchError>(Error::<T>::PendingInboundEntryNotFound.into())
+					}
+					Some(stored_inbound_entry) => {
+						let post_dispatch_entry = InboundEntry::create_post_voting_entry(
+							stored_inbound_entry,
+							session_id,
+							expected_proof_count,
+						)?;
+
+						*storage_entry = post_dispatch_entry;
+
+						Ok(())
+					}
+				}
+			})?;
+		}
+
+		Ok(())
+	}
+
+	/// Iterates over a batch of messages and checks if the requirements for
+	/// processing each message are met.
+	pub(crate) fn process_inbound_message(
```

Comment:
Added a TODO for me to find a way to factorize those `Err`, should there be some.

Reply:
Some evm crates have macros like:
```

macro_rules! try_or_fail {
	( $e:expr ) => {
		match $e {
			Ok(v) => v,
			Err(e) => return Control::Exit(e.into()),
		}
	};
}
```

Was thinking that we can do something similar.

Reply:
Interesting 😲.

Will think about this in a next PR with well-weight treatment 👍🏻

Reply:
Awesome, thank you!